File: Common/Material.cs
using System.Collections.Generic;
using OpenTK.Graphics.OpenGL4;

namespace Nebula;

public class Material {
    public Mesh[] meshes;
    public Texture[] textures;

    public Material(Mesh[] meshes, Texture[] textures) {
        this.meshes = meshes;
        this.textures = textures;
    }

    public void BindTexturesAndDraw(Shader shader) {
        for (int i = 0; i < textures.Length; i++) {
            textures[i].Bind(textureUnitLookup[i]);
            shader.SetUInt($"material.texture_{textures[i].type}", i);
        }

        foreach(Mesh mesh in meshes) mesh.Draw();
    }

    private readonly Dictionary<int, TextureUnit> textureUnitLookup = new Dictionary<int, TextureUnit>() {
        { 0, TextureUnit.Texture0 },
        { 1, TextureUnit.Texture1 },
        { 2, TextureUnit.Texture2 },
        { 3, TextureUnit.Texture3 },
        { 4, TextureUnit.Texture4 },
        { 5, TextureUnit.Texture5 },
        { 6, TextureUnit.Texture6 },
        { 7, TextureUnit.Texture7 },
        { 8, TextureUnit.Texture8 },
        { 9, TextureUnit.Texture9 }
    };
}

File: lib/common/material.dart
import 'package:nebula/nebula.dart';

import 'common.dart';

class Material {
  late final List<Mesh> meshes;
  late final List<Texture> textures;

  Material(this.meshes, this.textures);

  void bindTexturesAndDraw(Shader shader) {
    for (int i = 0; i < textures.length; i++) {
      textures[i].bind(textureUnitLookup[i]!);
      shader.setUInt('nebula_material.texture_${textures[i].type}', i);
    }

    draw();
  }

  void draw() {
    for (final mesh in meshes) {
      mesh.draw();
    }
  }
}

const textureUnitLookup = {
  0: GL_TEXTURE0,
  1: GL_TEXTURE1,
  2: GL_TEXTURE2,
  3: GL_TEXTURE3,
  4: GL_TEXTURE4,
  5: GL_TEXTURE5,
  6: GL_TEXTURE6,
  7: GL_TEXTURE7,
  8: GL_TEXTURE8,
  9: GL_TEXTURE9,
};
